src/langrila/memory/s3.py

```diff
@@ -0,0 +1,84 @@
+import os
+import logging
+import json
+from typing import Optional, Union
+
+import boto3
+from botocore.exceptions import ClientError
+from botocore.client import Config as BotoConfig
+
+from ..base import BaseConversationMemory
+
+class S3ConversationMemory(BaseConversationMemory):
+    def __init__(
+            self,
+            bucket: str,
+            key: str,
+            *,
+            region_name: Optional[str] = None,
+            api_version: Optional[str] = None,
+            use_ssl: Optional[bool] = True,
+            verify: Union[str, bool, None] = None,
+            endpoint_url: Optional[str] = None,
+            aws_access_key_id: Optional[str] = None,
+            aws_secret_access_key: Optional[str] = None,
+            aws_session_token: Optional[str] = None,
+            boto_config: Optional[BotoConfig] = None,
+        ):
+        self.bucket = bucket
+        self.key = key
+        self.region_name = region_name
+        self.api_version = api_version
+        self.use_ssl = use_ssl
+        self.verify = verify
+        self.endpoint_url = endpoint_url
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.aws_session_token = aws_session_token
+        self.boto_config = boto_config
+
+        self.s3_client = boto3.client(
+            's3',
+            region_name=self.region_name,
+            api_version=self.api_version,
+            use_ssl=self.use_ssl,
+            verify=self.verify,
+            endpoint_url=self.endpoint_url,
+            aws_access_key_id=self.aws_access_key_id,
+            aws_secret_access_key=self.aws_secret_access_key,
+            aws_session_token=self.aws_session_token,
+            config=self.boto_config
+        )
+
+        try:
+            # check at minimum
+            self.s3_client.head_bucket(Bucket=self.bucket)
+            logging.info(f"s3 connect success. bucket_name: {self.bucket}")
+        except Exception as e:
+            logging.error(f"s3 connect failed:{e}")
+            raise
+
+    def store(self, conversation_history: list[dict[str, str]]):
+        try:
+            json_data = json.dumps(conversation_history, ensure_ascii=False)
+            self.s3_client.put_object(Bucket=self.bucket, Key=self.key, Body=json_data)
+        except Exception as e:
+            logging.error(f"s3 store failed: {e}")
+            raise
+
+    def load(self) -> list[dict[str, str]]:
+        try:
```

Comment:
```suggestion
        if self.key is None:
            return []
        try:
```

It's needed if key is not given when instantiating this class.

Reply:
will fix.